src/store.cc

```diff
@@ -1285,6 +1304,7 @@ storeRegisterWithCacheManager(void)
     Mgr::RegisterAction("store_io", "Store IO Interface Stats", &Mgr::StoreIoAction::Create, 0, 1);
     Mgr::RegisterAction("store_check_cachable_stats", "storeCheckCachable() Stats",
                         storeCheckCachableStats, 0, 1);
+    Mgr::RegisterAction("store_queues", "SMP Transients and Caching Queues", StatQueues, 0, 1);
```

Comment:
Ideally this would be using an Action class to manage the report. See Mgr::StoreIoAction for example.

That would allow cleaner separation of the report logic into src/store/ or src/mgr/ and setting of the report Content-Type to "text/yaml" correctly.

Reply:
This Mgr::StoreIoAction (and other similar Actions) is a kind of 'aggregatable' Action. For example, such actions redefine their virtual collect() for this aggregation. On the other hand, StatQueues is not aggregatable and that is why we should use the existing FunAction for it (which is selected in our case).

Reply:
> See Mgr::StoreIoAction for example

I would not use that class as an example. Like most class-based Actions, StoreIoAction is aggregatable() while the new report is not (and should not be). If we somehow decide that duplicating code is a good idea, then we should duplicate Mgr::FunAction instead, calling the reporting method instead of the OBJH handler.

> That would allow cleaner separation of the report logic into src/store/ or src/mgr/

AFAICT, the separation of Queue reporting methods from the Queue classes has nothing to do with the reporting methods (not) being Action class methods: We can put Queue reporting functions into src/store/ or src/mgr/, without making them Action class methods.

There is at least one serious reason to reject moving Queue reporting functions away from Queue: This PR places Queue reporting together with Queue implementation to allow the intricate reporting code to safely access sensitive Queue data members (in shared memory) without the overheads/risks of exposing those private members. Having that safe/convenient access is especially important here because there is not just one Queue class but several layers of classes that the reporting code has to navigate. @yadij, what different placement principle do you suggest that would overwrite the current "convenience/safety of access" placement principle?

> setting of the report Content-Type to "text/yaml" correctly

That would be a mistake: While individual kid reports are valid YAML, the response itself is not. We need to change how Squid wraps/separates individual kid responses. After that, we can start using text/yaml for valid YAML responses. That global enhancement is not trivial and is outside this project scope. IMO, it should be done when adding proper YAML generation support -- a stand-alone project!

@yadij, do the above arguments against using an Action class address you change request?

Reply:
My understanding of Action is that we are trying to move the cachemgr API over to Actions and FunAction is a wrapper for legacy C code.

Yes StoreIoAction may not have been the best example. This new report would initially only need to override Action::dump() with the logic currently inside the StatQueues() function being registered.

If the current output is not validating as YAML properly it should be made to do so. Using standard formats properly instead of in-house custom plain-text is the "good design practice" I am aiming at here. I don't particularly care *which* standard format right now since we do not have a formal policy on that.

Reply:
> This new report would initially only need to override Action::dump()

I am afraid that this is insufficient because our action is not aggregatable, i.e., it sends the response directly to HTTP client. In FunAction, this logic is implemented in Mgr::FunAction::respond(), where the work is delegated to Mgr::Filler() and Mgr::StoreToCommWriter(). I am not sure that it is a good idea to copy and paste this logic into a new hypothetical class. We could probably refactor this area, implementing a new base class for non-aggregatable actions so that the descendants would only need overriding their 'dump()' as you noted. But I think this refactoring is out of scope - it should be done as a separate work.

Reply:
> My understanding of Action is that we are trying to move the cachemgr API over to Actions and FunAction is a wrapper for legacy C code

The above understanding does not reflect the current reality well. FunAction is a wrapper for non-aggregating function-initiated reports. Yes, FunAction is used for legacy reports that were first implemented using C functions, but bare functions are not second class citizens in C++. If something can be equally well implemented using a function and a class, the former should be preferred because function is a simpler concept. There is nothing wrong with using functions in modern C++, of course! This PR is using a simple StatQueues() function to jump-start queue reports. I see no problems with that solution. Creating a whole new class just to wrap those eight lines of code would make things worse, not better!

The existing function-vs-class separation between FunAction and aggregating Actions is a bit of a red herring: I am pretty sure that some of those aggregating Actions should have been implemented using a simple function as well (with a common Action-based class that does the rest). There is some code duplication among those aggregating Actions. The initial refactoring solved many legacy problems and added SMP support, but we probably missed some good opportunities. Such defects are common for complex cleanup tasks, especially when they do not go through proper reviews. Things are clearer now, and we may do another refactoring pass eventually.

> If the current output is not validating as YAML properly it should be made to do so.

Yes, but not here/now. None of the non-aggregating reports are validating as YAML today (because of how kid-specific reports are separated/wrapped). Fixing that old problem is outside this project scope because the fix affects _existing_ reports (and automation that uses them) and will benefit from being combined with other YAML-related improvements. What is in scope for this (and any similar future) PR is generating YAML-compliant per-kid output. This PR does that.

Reply:
> I don't particularly care which standard format right now since we do not have a formal policy on that.

This is an out of scope discussion, but my review of the old squid-dev threads and improved understanding of YAML shows that we have only two viable options left:

1. YAML (with formatting manipulators ensuring the right syntax).
2. Configurable output format (including YAML) based on the generic formatting manipulators [ideas](http://www.squid-cache.org/mail-archive/squid-dev/201404/0009.html).

Thus, until we have the right formatting manipulators, new aggregated reports and new kid-specific parts of non-aggregated reports should use YAML. This PR follows that principle.

----

* http://lists.squid-cache.org/pipermail/squid-dev/2015-August/002953.html
* http://www.squid-cache.org/mail-archive/squid-dev/201404/0009.html
* http://www.squid-cache.org/mail-archive/squid-dev/201110/0309.html
* http://www.squid-cache.org/mail-archive/squid-dev/201101/0146.html

Reply:
@rousskov, I prefer to use YAML yes - due to its relatively human-readable nature. But reports are not restricted to that. Any class inheriting from Action can override contentType() to deliver other standardized formats.

@eduard-bagdasaryan, the class Action API has two sections - collate() + respond() are part of the per-worker API, and dump() + contentType() which I am suggesting be used are for non-aggregation reports like this one.

Reply:
> FunAction restricts the report to text/plain 

which is what the new report is. Again, the new report (as a whole) is not YAML and cannot be YAML (within this PR scope). Future refactoring can make function-initiated reports, like this one, to produce YAML _without_ converting them into classes.

> and require a refactoring when the class is made to send a non-custom format.

That is a speculation, not a fact. I can easily imagine scenarios where no such refactoring would be necessary or, to be more precise, where the amount of refactoring would be the same (or smaller) if this report is initiated by a function rather than a class. Again, you are basing your suggestion on a false premise that a class would be better than the function here. In this context, the reverse is true. A function is the simplest and overall best solution here.

> This being a brand new report (debug feature) I would like to see it polished up to fully before being added.

Your request is not a polishing request. The changes you want will make the code _worse_, not better. Adding a class just to convert a function into a method would result in more code duplication, increased complexity, more special cases, more ink, etc. All the given reasons for adding a new class were either factually wrong (e.g., "the new class will return responses with YAML content type") or unsubstantiated speculation where other experts see a different picture (i.e. "it will require less refactoring in the future"). If I missed any valid reason where a class would work better than the function, please let me know.

> the class Action API has two sections - collate() + respond() are part of the per-worker API, and dump() + contentType() [...] for non-aggregation reports like this one.

There are no such "sections" in the Action API. The collect() method is not paired with respond(), contentType() is not paired with dump(), respond() is used for all reports, etc. The API is fairly complex because it needs to solve several non-trivial problems for two partially overlapping categories of reports in SMP and non-SMP environments. I am sure it can be improved as well -- as discussed earlier, it probably mistreats function-based report initiators (an API mistake that probably contributes to misunderstandings like the ones expressed on this change request thread).

However, all of that complexity should not really matter here. A simple initiation function is the best (and already implemented!) solution for this report. It is simple, it follows a common well-known pattern, it produces the best result possible today, it does not duplicate code, it is unlikely to increase future refactoring costs. Converting it into a new class will have the opposite effect on most of these categories.